File: flutter/packages/workern_models/lib/attribution.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:workern_models/base.dart';
import 'package:workern_utils/workern_utils.dart';

enum DeviceType { DESKTOP, MOBILE, TABLET, BOT, UNKNOWN }

class AttributionClick {
  final DateTime timestamp;
  final String ipAddress;
  final String ipHash;
  final String userAgent;
  final DeviceType deviceType;
  final bool isBot;
  final String? referrer;
  final String? country;
  final String? city;
  final String? fingerprint;

  AttributionClick({
    required this.timestamp,
    required this.ipAddress,
    required this.ipHash,
    required this.userAgent,
    required this.deviceType,
    required this.isBot,
    this.referrer,
    this.country,
    this.city,
    this.fingerprint,
  });

  factory AttributionClick.fromJson(Map<String, dynamic> json) {
    return AttributionClick(
      timestamp: parseTimestamp(json['timestamp']) ?? DateTime.now(),
      ipAddress: json['ipAddress'] ?? '',
      ipHash: json['ipHash'] ?? '',
      userAgent: json['userAgent'] ?? '',
      deviceType: DeviceType.values.firstWhere(
        (e) => e.name == json['deviceType'],
        orElse: () => DeviceType.UNKNOWN,
      ),
      isBot: json['isBot'] ?? false,
      referrer: json['referrer'],
      country: json['country'],
      city: json['city'],
      fingerprint: json['fingerprint'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'timestamp': timestamp,
      'ipAddress': ipAddress,
      'ipHash': ipHash,
      'userAgent': userAgent,
      'deviceType': deviceType.name,
      'isBot': isBot,
      'referrer': referrer,
      'country': country,
      'city': city,
      'fingerprint': fingerprint,
    };
  }
}

class AttributionLink extends Base {
  // Link Details
  final String shortCode;
  final String originalUrl;
  final String fullShortUrl;

  // Context
  final String? campaignId;
  final String? submissionId;
  final String? creatorId;

  // Tracking
  final int totalClicks;
  final int uniqueClicks;
  final int botClicks;

  // Analytics
  final DateTime? lastClickedAt;
  final Map<String, int> clicksByCountry;
  final Map<String, int> clicksByDevice;

  // Status
  final bool isActive;
  final DateTime? expiresAt;

  AttributionLink({
    required super.id,
    required super.createdAt,
    required super.updatedAt,
    required super.owner,
    required super.space,
    required this.shortCode,
    required this.originalUrl,
    required this.fullShortUrl,
    this.campaignId,
    this.submissionId,
    this.creatorId,
    required this.totalClicks,
    required this.uniqueClicks,
    required this.botClicks,
    this.lastClickedAt,
    required this.clicksByCountry,
    required this.clicksByDevice,
    required this.isActive,
    this.expiresAt,
  });

  factory AttributionLink.fromJson(Map<String, dynamic> json) {
    return AttributionLink(
      id: json['id'] ?? '',
      createdAt: parseTimestamp(json['createdAt']) ?? DateTime.now(),
      updatedAt: parseTimestamp(json['updatedAt']) ?? DateTime.now(),
      owner: BaseOwner.fromJson(json['owner'] ?? {}),
      space: BaseSpace.fromJson(json['space'] ?? {}),
      shortCode: json['shortCode'] ?? '',
      originalUrl: json['originalUrl'] ?? '',
      fullShortUrl: json['fullShortUrl'] ?? '',
      campaignId: json['campaignId'],
      submissionId: json['submissionId'],
      creatorId: json['creatorId'],
      totalClicks: json['totalClicks'] ?? 0,
      uniqueClicks: json['uniqueClicks'] ?? 0,
      botClicks: json['botClicks'] ?? 0,
      lastClickedAt: json['lastClickedAt'] != null
          ? parseTimestamp(json['lastClickedAt'])
          : null,
      clicksByCountry: Map<String, int>.from(json['clicksByCountry'] ?? {}),
      clicksByDevice: Map<String, int>.from(json['clicksByDevice'] ?? {}),
      isActive: json['isActive'] ?? true,
      expiresAt: json['expiresAt'] != null
          ? parseTimestamp(json['expiresAt'])
          : null,
    );
  }

  @override
  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'createdAt': createdAt,
      'updatedAt': updatedAt,
      'owner': owner.toJson(),
      'space': space.toJson(),
      'shortCode': shortCode,
      'originalUrl': originalUrl,
      'fullShortUrl': fullShortUrl,
      'campaignId': campaignId,
      'submissionId': submissionId,
      'creatorId': creatorId,
      'totalClicks': totalClicks,
      'uniqueClicks': uniqueClicks,
      'botClicks': botClicks,
      'lastClickedAt': lastClickedAt,
      'clicksByCountry': clicksByCountry,
      'clicksByDevice': clicksByDevice,
      'isActive': isActive,
      'expiresAt': expiresAt,
    };
  }
}

class AttributionAnalytics {
  final String linkId;
  final int totalClicks;
  final int uniqueClicks;
  final int botClicks;
  final double conversionRate;
  final List<CountryClicks> topCountries;
  final List<DeviceClicks> topDevices;
  final List<ClicksOverTime> clicksOverTime;

  AttributionAnalytics({
    required this.linkId,
    required this.totalClicks,
    required this.uniqueClicks,
    required this.botClicks,
    required this.conversionRate,
    required this.topCountries,
    required this.topDevices,
    required this.clicksOverTime,
  });

  factory AttributionAnalytics.fromJson(Map<String, dynamic> json) {
    return AttributionAnalytics(
      linkId: json['linkId'] ?? '',
      totalClicks: json['totalClicks'] ?? 0,
      uniqueClicks: json['uniqueClicks'] ?? 0,
      botClicks: json['botClicks'] ?? 0,
      conversionRate: (json['conversionRate'] as num?)?.toDouble() ?? 0,
      topCountries:
          (json['topCountries'] as List?)
              ?.map((e) => CountryClicks.fromJson(e))
              .toList() ??
          [],
      topDevices:
          (json['topDevices'] as List?)
              ?.map((e) => DeviceClicks.fromJson(e))
              .toList() ??
          [],
      clicksOverTime:
          (json['clicksOverTime'] as List?)
              ?.map((e) => ClicksOverTime.fromJson(e))
              .toList() ??
          [],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'linkId': linkId,
      'totalClicks': totalClicks,
      'uniqueClicks': uniqueClicks,
      'botClicks': botClicks,
      'conversionRate': conversionRate,
      'topCountries': topCountries.map((e) => e.toJson()).toList(),
      'topDevices': topDevices.map((e) => e.toJson()).toList(),
      'clicksOverTime': clicksOverTime.map((e) => e.toJson()).toList(),
    };
  }
}

class CountryClicks {
  final String country;
  final int clicks;

  CountryClicks({required this.country, required this.clicks});

  factory CountryClicks.fromJson(Map<String, dynamic> json) {
    return CountryClicks(
      country: json['country'] ?? '',
      clicks: json['clicks'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {'country': country, 'clicks': clicks};
  }
}

class DeviceClicks {
  final String device;
  final int clicks;

  DeviceClicks({required this.device, required this.clicks});

  factory DeviceClicks.fromJson(Map<String, dynamic> json) {
    return DeviceClicks(
      device: json['device'] ?? '',
      clicks: json['clicks'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {'device': device, 'clicks': clicks};
  }
}

class ClicksOverTime {
  final String date;
  final int clicks;

  ClicksOverTime({required this.date, required this.clicks});

  factory ClicksOverTime.fromJson(Map<String, dynamic> json) {
    return ClicksOverTime(
      date: json['date'] ?? '',
      clicks: json['clicks'] ?? 0,
    );
  }

  Map<String, dynamic> toJson() {
    return {'date': date, 'clicks': clicks};
  }
}

File: tools/firebase/functions-ts/src/tracking/attribution.ts
/**
 * Attribution Link Tracking System
 *
 * Enterprise-grade link attribution tracking with:
 * - Bot detection using user-agent analysis
 * - IP-based deduplication with hashing for privacy
 * - Browser fingerprinting for enhanced deduplication
 * - Geographic tracking (country/city)
 * - Device type detection
 * - Real-time analytics
 */

import { onRequest } from 'firebase-functions/v2/https';
import { db, firestoreWriteTimestamp } from '../global';
import * as crypto from 'crypto';
import * as admin from 'firebase-admin';

// Bot detection patterns (inspired by Google Analytics, Cloudflare, etc.)
const BOT_PATTERNS = [
  /bot/i,
  /crawler/i,
  /spider/i,
  /scraper/i,
  /curl/i,
  /wget/i,
  /python-requests/i,
  /java/i,
  /okhttp/i,
  /axios/i,
  /fetch/i,
  /lighthouse/i,
  /googlebot/i,
  /bingbot/i,
  /slackbot/i,
  /facebookexternalhit/i,
  /twitterbot/i,
  /linkedinbot/i,
  /whatsapp/i,
  /telegrambot/i
];

/**
 * Detect if request is from a bot
 */
function isBot(userAgent: string): boolean {
  return BOT_PATTERNS.some((pattern) => pattern.test(userAgent));
}

/**
 * Detect device type from user agent
 */
function getDeviceType(
  userAgent: string
): 'DESKTOP' | 'MOBILE' | 'TABLET' | 'BOT' | 'UNKNOWN' {
  if (isBot(userAgent)) return 'BOT';

  const ua = userAgent.toLowerCase();

  if (ua.includes('tablet') || ua.includes('ipad')) return 'TABLET';
  if (ua.includes('mobile') || ua.includes('android') || ua.includes('iphone'))
    return 'MOBILE';
  if (
    ua.includes('windows') ||
    ua.includes('macintosh') ||
    ua.includes('linux')
  )
    return 'DESKTOP';

  return 'UNKNOWN';
}

/**
 * Hash IP address for privacy compliance (GDPR, CCPA)
 */
function hashIP(ip: string): string {
  return crypto.createHash('sha256').update(ip).digest('hex');
}

/**
 * Generate browser fingerprint from available data
 */
function generateFingerprint(
  userAgent: string,
  acceptLanguage?: string
): string {
  const data = `${userAgent}:${acceptLanguage || 'unknown'}`;
  return crypto.createHash('md5').update(data).digest('hex');
}

/**
 * Generate short code for attribution link
 */
function generateShortCode(): string {
  const chars =
    'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  let result = '';
  for (let i = 0; i < 8; i++) {
    result += chars.charAt(Math.floor(Math.random() * chars.length));
  }
  return result;
}

/**
 * Check if click should be counted (deduplication)
 */
async function shouldCountClick(
  linkId: string,
  ipHash: string,
  fingerprint: string
): Promise<boolean> {
  // Check if IP+fingerprint combination clicked in last 24 hours
  const yesterday = new Date(Date.now() - 24 * 60 * 60 * 1000);

  const recentClicks = await db
    .collection('attributionClicks')
    .where('linkId', '==', linkId)
    .where('ipHash', '==', ipHash)
    .where('fingerprint', '==', fingerprint)
    .where('timestamp', '>', yesterday)
    .limit(1)
    .get();

  return recentClicks.empty;
}

/**
 * Firebase Function: Create Attribution Link
 * Callable function to create a new attribution/tracking link
 */
export const createAttributionLink = onRequest(
  { cors: true, region: 'asia-south1' },
  async (req, res) => {
    try {
      const {
        originalUrl,
        campaignId,
        submissionId,
        creatorId,
        expiresAt,
        userId,
        userDisplayName
      } = req.body;

      if (!originalUrl || !userId) {
        res.status(400).json({ error: 'originalUrl and userId are required' });
        return;
      }

      const shortCode = generateShortCode();
      const fullShortUrl = `https://workern.app/l/${shortCode}`;

      const linkData = {
        id: shortCode,
        shortCode,
        originalUrl,
        fullShortUrl,
        campaignId: campaignId || null,
        submissionId: submissionId || null,
        creatorId: creatorId || null,
        totalClicks: 0,
        uniqueClicks: 0,
        botClicks: 0,
        lastClickedAt: null,
        clicksByCountry: {},
        clicksByDevice: {},
        isActive: true,
        expiresAt: expiresAt ? new Date(expiresAt) : null,
        createdAt: firestoreWriteTimestamp,
        updatedAt: firestoreWriteTimestamp,
        owner: { uid: userId, name: userDisplayName || 'Unknown' },
        space: { id: 'attributionApp' }
      };

      // Save to Firestore (global path + user path)
      const batch = db.batch();

      // Global path
      const globalRef = db
        .collection('apps/attributionApp/links')
        .doc(shortCode);
      batch.set(globalRef, linkData);

      // User path
      const userRef = db
        .collection(`users/${userId}/mySpaces/attributionApp/links`)
        .doc(shortCode);
      batch.set(userRef, linkData);

      await batch.commit();

      res.json({
        success: true,
        link: linkData
      });
    } catch (error) {
      console.error('Error creating attribution link:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  }
);

/**
 * Firebase Function: Track Click (Redirect)
 * HTTP endpoint that tracks click and redirects to original URL
 */
export const trackClick = onRequest(
  { cors: true, region: 'asia-south1' },
  async (req, res) => {
    try {
      const shortCode = req.path.replace('/l/', '').replace('/', '');

      if (!shortCode) {
        res.status(400).send('Invalid link');
        return;
      }

      // Get link from Firestore
      const linkDoc = await db
        .collection('apps/attributionApp/links')
        .doc(shortCode)
        .get();

      if (!linkDoc.exists) {
        res.status(404).send('Link not found');
        return;
      }

      const linkData = linkDoc.data();

      if (!linkData?.isActive) {
        res.status(410).send('Link expired or inactive');
        return;
      }

      if (linkData.expiresAt && new Date(linkData.expiresAt) < new Date()) {
        res.status(410).send('Link expired');
        return;
      }

      // Extract request data
      const ipAddress = req.ip || req.headers['x-forwarded-for'] || 'unknown';
      const userAgent = req.headers['user-agent'] || 'unknown';
      const acceptLanguage = req.headers['accept-language'] || undefined;
      const referrer =
        req.headers['referer'] || req.headers['referrer'] || undefined;

      const ipHash = hashIP(String(ipAddress));
      const fingerprint = generateFingerprint(
        String(userAgent),
        String(acceptLanguage)
      );
      const deviceType = getDeviceType(String(userAgent));
      const isBotRequest = isBot(String(userAgent));

      // Check if click should be counted (deduplication)
      const isUniqueClick = await shouldCountClick(
        shortCode,
        ipHash,
        fingerprint
      );

      // Save click data
      const clickData = {
        linkId: shortCode,
        timestamp: firestoreWriteTimestamp,
        ipAddress: String(ipAddress),
        ipHash,
        userAgent: String(userAgent),
        deviceType,
        isBot: isBotRequest,
        referrer,
        fingerprint,
        counted: !isBotRequest && isUniqueClick
      };

      // Save click asynchronously (don't block redirect)
      db.collection('attributionClicks').add(clickData);

      // Update link analytics
      const updateData: any = {
        totalClicks: admin.firestore.FieldValue.increment(1),
        lastClickedAt: firestoreWriteTimestamp,
        updatedAt: firestoreWriteTimestamp
      };

      if (isBotRequest) {
        updateData.botClicks = admin.firestore.FieldValue.increment(1);
      } else if (isUniqueClick) {
        updateData.uniqueClicks = admin.firestore.FieldValue.increment(1);
      }

      // Update country clicks
      updateData[`clicksByDevice.${deviceType}`] =
        admin.firestore.FieldValue.increment(1);

      // Update both global and user paths
      const batch = db.batch();
      batch.update(linkDoc.ref, updateData);

      if (linkData.owner?.uid) {
        const userLinkRef = db
          .collection(
            `users/${linkData.owner.uid}/mySpaces/attributionApp/links`
          )
          .doc(shortCode);
        batch.update(userLinkRef, updateData);
      }

      batch.commit().catch(console.error);

      // Redirect to original URL
      res.redirect(302, linkData.originalUrl);
    } catch (error) {
      console.error('Error tracking click:', error);
      res.status(500).send('Internal server error');
    }
  }
);

/**
 * Firebase Function: Get Analytics
 * Callable function to retrieve analytics for an attribution link
 */
export const getAttributionAnalytics = onRequest(
  { cors: true, region: 'asia-south1' },
  async (req, res) => {
    try {
      const { linkId } = req.body;

      if (!linkId) {
        res.status(400).json({ error: 'linkId is required' });
        return;
      }

      const linkDoc = await db
        .collection('apps/attributionApp/links')
        .doc(linkId)
        .get();

      if (!linkDoc.exists) {
        res.status(404).json({ error: 'Link not found' });
        return;
      }

      const linkData = linkDoc.data();

      // Get recent clicks for time-series data
      const sevenDaysAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000);
      const recentClicks = await db
        .collection('attributionClicks')
        .where('linkId', '==', linkId)
        .where('timestamp', '>', sevenDaysAgo)
        .where('isBot', '==', false)
        .get();

      // Aggregate data
      const clicksByDate: { [date: string]: number } = {};
      const clicksByCountry: { [country: string]: number } = {};

      recentClicks.forEach((doc) => {
        const click = doc.data();
        const date = new Date(click.timestamp._seconds * 1000)
          .toISOString()
          .split('T')[0];

        clicksByDate[date] = (clicksByDate[date] || 0) + 1;

        if (click.country) {
          clicksByCountry[click.country] =
            (clicksByCountry[click.country] || 0) + 1;
        }
      });

      const analytics = {
        linkId,
        totalClicks: linkData?.totalClicks || 0,
        uniqueClicks: linkData?.uniqueClicks || 0,
        botClicks: linkData?.botClicks || 0,
        conversionRate: 0, // Would need conversion tracking
        topCountries: Object.entries(clicksByCountry)
          .map(([country, clicks]) => ({ country, clicks }))
          .sort((a, b) => b.clicks - a.clicks)
          .slice(0, 10),
        topDevices: Object.entries(linkData?.clicksByDevice || {})
          .map(([device, clicks]) => ({ device, clicks }))
          .sort((a, b) => (b.clicks as number) - (a.clicks as number)),
        clicksOverTime: Object.entries(clicksByDate)
          .map(([date, clicks]) => ({ date, clicks }))
          .sort((a, b) => a.date.localeCompare(b.date))
      };

      res.json(analytics);
    } catch (error) {
      console.error('Error getting analytics:', error);
      res.status(500).json({ error: 'Internal server error' });
    }
  }
);
